The `timeout` parameter is only applied to the assertion part of the step

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
If the matter is security sensitive, please disclose it privately to a security contact: https://github.com/kudobuilder/kuttl/blob/master/SECURITY_CONTACTS
-->

**What happened**:
While working on a test that includes a webhook, I noticed that `timeout` parameter is not used when the test step fails in the apply-phase:

```
    harness.go:66: going to run test suite with timeout of 3000000 seconds for each step
    --- FAIL: kudo/harness (0.02s)
        --- FAIL: kudo/harness/terminal-failed-job (7.67s)
            logger.go:37: 14:51:05 | terminal-failed-job | Ignoring failjob-operator as it does not match file name regexp: ^(\d+)-([^.]+)(.yaml)?$
            logger.go:37: 14:51:05 | terminal-failed-job | Ignoring job-timeout-operator as it does not match file name regexp: ^(\d+)-([^.]+)(.yaml)?$
            logger.go:37: 14:51:05 | terminal-failed-job | Creating namespace: kudo-test-heroic-dodo
            logger.go:37: 14:51:05 | terminal-failed-job/0-install | starting test step 0-install
            logger.go:37: 14:51:05 | terminal-failed-job/0-install | running command:  {kubectl kudo install --instance failjob-operator-instance ./failjob-operator %!s(bool=true) %!s(bool=false) %!s(bool=false)}
            logger.go:37: 14:51:12 | terminal-failed-job/0-install | Error: failed to install instance failjob-operator-instance: installing Instance: Internal error occurred: failed calling webhook "instance-admission.kudo.dev": Post https://kudo-controller-manager-service.kudo-system.svc:443/admit-kudo-dev-v1beta1-instance?timeout=30s: dial tcp 10.103.201.124:443: connect: connection refused

            logger.go:37: 14:51:12 | terminal-failed-job/0-install | operator.kudo.dev/v1beta1/failjob-operator created
                operatorversion.kudo.dev/v1beta1/failjob-operator-0.1.0 created

            case.go:156: exit status 255
            case.go:159: failed in step 0-install
```

The [code suggests](https://github.com/kudobuilder/kuttl/blob/master/pkg/test/step.go#L393) that the `timeout` is only respected during the assertion phase of the step. I  often use the timeout to "pause" the execution of the failing step and look into the cluster (using `kubectl`). A number of k8s admission controllers might reject applied resources so it might be beneficial to extend timeout to the whole step and not just the assertion phase.

**What you expected to happen**:
The `timeout` should be applied to the whole step.

**How to reproduce it (as minimally and precisely as possible)**:
Using any step command that fails (e.g. `false`)

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- KUTTL version (use `kubectl kuttl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/test/case.go
+++ b/pkg/test/case.go
@@ -192,10 +192,10 @@ func (t *Case) Run(test *testing.T, tc *report.Testcase) {
 			caseErr := fmt.Errorf("failed in step %s", testStep.String())
 			tc.Failure = report.NewFailure(caseErr.Error(), errs)
 
+			test.Error(caseErr)
 			for _, err := range errs {
 				test.Error(err)
 			}
-			test.Error(caseErr)
 			break
 		}
 	}

--- a/pkg/test/step.go
+++ b/pkg/test/step.go
@@ -174,8 +174,15 @@ func (s *Step) Create(namespace string) []error {
 			errors = append(errors, err)
 			continue
 		}
+		ctx := context.Background()
+		if s.Timeout > 0 {
+			var cancel context.CancelFunc
+			//TODO (kensipe): fix time s.Timeout
+			ctx, cancel = context.WithTimeout(ctx, time.Duration(2)*time.Second)
+			defer cancel()
+		}
 
-		if updated, err := testutils.CreateOrUpdate(context.TODO(), cl, obj, true); err != nil {
+		if updated, err := testutils.CreateOrUpdate(ctx, cl, obj, true); err != nil {
 			errors = append(errors, err)
 		} else {
 			action := "created"

--- a/pkg/test/step.go
+++ b/pkg/test/step.go
@@ -177,8 +177,7 @@ func (s *Step) Create(namespace string) []error {
 		ctx := context.Background()
 		if s.Timeout > 0 {
 			var cancel context.CancelFunc
-			//TODO (kensipe): fix time s.Timeout
-			ctx, cancel = context.WithTimeout(ctx, time.Duration(2)*time.Second)
+			ctx, cancel = context.WithTimeout(ctx, time.Duration(s.Timeout)*time.Second)
 			defer cancel()
 		}
 

--- a/pkg/test/utils/kubernetes.go
+++ b/pkg/test/utils/kubernetes.go
@@ -767,8 +767,7 @@ func CreateOrUpdate(ctx context.Context, cl client.Client, obj runtime.Object, r
 	if retryOnError {
 		validators = append(validators, k8serrors.IsConflict)
 	}
-
-	return updated, Retry(ctx, func(ctx context.Context) error {
+	err = Retry(ctx, func(ctx context.Context) error {
 		expected := orig.DeepCopyObject()
 		actual := orig.DeepCopyObject()
 
@@ -792,6 +791,10 @@ func CreateOrUpdate(ctx context.Context, cl client.Client, obj runtime.Object, r
 		}
 		return err
 	}, validators...)
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+		err = errors.New("create/update timeout exceeded")
+	}
+	return updated, err
 }
 
 // SetAnnotation sets the given key and value in the object's annotations, returning a copy.

